packages/@aws-cdk/aws-iot-actions/lib/kinesis-put-record-action.ts

```diff
@@ -0,0 +1,54 @@
+import * as iam from '@aws-cdk/aws-iam';
+import * as iot from '@aws-cdk/aws-iot';
+import * as kinesis from '@aws-cdk/aws-kinesis';
+import { CommonActionProps } from './common-action-props';
+import { singletonActionRole } from './private/role';
+
+/**
+ * Configuration properties of an action for the Kinesis Data stream.
+ */
+export interface KinesisPutRecordActionProps extends CommonActionProps {
+  /**
+   * The partition key used to determine to which shard the data is written.
+   * The partition key is usually composed of an expression (for example, ${topic()} or ${timestamp()}).
+   * For more information @see https://docs.aws.amazon.com/kinesis/latest/APIReference/API_PutRecord.html#API_PutRecord_RequestParameters
+   *
+   * @default - None
```

Comment:
Can you elaborate what does "None" mean here?

Reply:
Sorry, I tried to elaborate, but I could not😞.

Proberbly, IoT Core rule is fill default value (e.g. MQTT payload JSON string) because `PartitionKey` is required in [this API Reference](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_PutRecord.html#API_PutRecord_RequestParameters).
But there is no description about it in [IoT Core rule documentation](https://docs.aws.amazon.com/iot/latest/developerguide/kinesis-rule-action.html#kinesis-rule-action-parameters) and [CloudFormation Documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iot-topicrule-kinesisaction.html#aws-properties-iot-topicrule-kinesisaction-properties)

Reply:
Did you try deploying an IoT app without this property filled?

Maybe this is simply a mistake in the CloudFormation resource, and this should actually be required?

Reply:
> Did you try deploying an IoT app without this property filled?

Yes I tried and the deploying secceeded.

Again, I deployed, and I tested following for confirmation to shard spreading:

- Prepare
  - Deploy Kinesis Stream with 10 shards
- Test1
  - Prepare: Deploy IoT Core Kinesis Rule action with `${timestamp()}` partitionKey
  - Send 10 same payloads
- Test2
  - Prepare: Deploy IoT Core Kinesis Rule action with no partitionKey
  - Send 10 same payloads
- Test3
  - Prepare: Deploy IoT Core Kinesis Rule action with no partitionKey
  - Send 10 payloads different each time

Results:

- Test1: Records was spreaded to each shards
- Test2: All records was put in same record
- Test3: All records was put in same record

The IoT Core may be filling in a static `partitionKey` if it is not specified.
But this is just a guess...

Reply:
OK. I think that's a good guess. Let's put that in the docs - it's certainly better than just "None", which gives you no clue what that actually means 🙂.

Thanks for the detailed testing!

Reply:
I don't hate it, but I'm a little concerned that the L2 won't have all of the capabilities of the underlying service (in this case, leaving the partition key as empty).

Does the console allow you to leave it as empty, or do you _have to_ provide a value?

Reply:
> the L2 won't have all of the capabilities of the underlying service

It is so reasonable I think! So I will never mind to revert my commit that add default value.

> Does the Console allow you to leave it as empty, or do you have to provide a value?

The Console requre PartitionKey, but [CloudFormation does not](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iot-topicrule-kinesisaction.html#:~:text=The%20partition%20key.-,Required%3A%20No,-Type%3A%20String). For keeping compatibility for CloudFormation, let's not add a default value, instead add a detailed description in JSDoc and Readme?

Reply:
Let's do this.

Make this field required, like it is in the console. But, allow specifying an empty string there (`''`). If an empty string is passed, then we will actually pass `undefined` to the underlying CloudFormation resource for that field. Of course, we will need to cover that in our documentation.

This way, we will be close to the Console experience, while covering the entire service catapbilities.

What do you think about this @yamatatsu?

Reply:
It's good I think! 👍🏻 
OK, I will do it!